flytepropeller/pkg/controller/nodes/array/handler.go

```diff
@@ -631,6 +655,46 @@ func (a *arrayNodeHandler) Handle(ctx context.Context, nCtx interfaces.NodeExecu
 	return handler.DoTransition(handler.TransitionTypeEphemeral, handler.PhaseInfoRunning(&handler.ExecutionInfo{})), nil
 }
 
+// OffloadLargeLiteral offloads the large literal if meets the threshold conditions
+func (a *arrayNodeHandler) OffloadLargeLiteral(ctx context.Context, datastore *storage.DataStore, dataReference storage.DataReference, toBeOffloaded *idlcore.Literal) error {
+	literalSizeBytes := uint64(proto.Size(toBeOffloaded))
+	literalSizeMB := literalSizeBytes / MB
+	// check if the literal is large
+	if literalSizeMB > a.literalOffloadingConfig.MaxSizeInMBForOffloading {
+		errString := fmt.Sprintf("Literal size [%d] MB is larger than the max size [%d] MB for offloading", literalSizeMB, a.literalOffloadingConfig.MaxSizeInMBForOffloading)
+		logger.Errorf(ctx, errString)
+		return fmt.Errorf(errString)
+	}
+	if literalSizeMB < a.literalOffloadingConfig.MinSizeInMBForOffloading {
+		logger.Debugf(ctx, "Literal size [%d] MB is smaller than the min size [%d] MB for offloading", literalSizeMB, a.literalOffloadingConfig.MinSizeInMBForOffloading)
+		return nil
+	}
+
+	inferredType := validators.LiteralTypeForLiteral(toBeOffloaded)
+	if inferredType == nil {
+		errString := "Failed to determine literal type for offloaded literal"
+		logger.Errorf(ctx, errString)
+		return fmt.Errorf(errString)
+	}
+
+	// offload the literal
+	if err := datastore.WriteProtobuf(ctx, dataReference, storage.Options{}, toBeOffloaded); err != nil {
+		logger.Errorf(ctx, "Failed to offload literal at location [%s] with error [%s]", dataReference, err)
+		return err
+	}
+
+	// update the literal with the offloaded URI, size and inferred type
+	toBeOffloaded.Value = &idlcore.Literal_OffloadedMetadata{
```

Comment:
nit: might be nice to debug log here that we offloaded and dump the uri?

Reply:
Added